File: java/146_LRU_Cache.java
// LinkedHashMap
class LRUCache extends LinkedHashMap<Integer,Integer>{
    private int capacity;

    public LRUCache(int capacity) {
        super(capacity,0.75F,true);
        this.capacity = capacity;
    }
    
    public int get(int key) {
        return super.getOrDefault(key,-1);
    }
    
    public void put(int key, int value) {
        super.put(key,value);
    }

    @Override
    protected boolean removeEldestEntry(Map.Entry<Integer,Integer> eldest){
        return size() > capacity;
    }
}

// HashMap + DoubleLinkedNode
class LRUCache {
    public class Node{
        int key;
        int value;
        Node prev;
        Node next;
    }

    private void addNode(Node node){
        node.prev = head;
        node.next = head.next;
        head.next.prev = node;
        head.next = node;
    } 

    private void removeNode(Node node){
        Node pre = node.prev;
        Node nxt = node.next;
        pre.next = nxt;
        nxt.prev = pre;
    }

    private void moveToHead(Node node){
        removeNode(node);
        addNode(node);
    }

    private Node deleteTail(){
        Node realTail = tail.prev;
        removeNode(realTail);
        return realTail;
    }

    Map<Integer, Node> cache = new HashMap<>();
    int curSize;
    int capacity;
    Node head;
    Node tail;
    
    public LRUCache(int capacity) {
        this.curSize = 0;
        this.capacity = capacity;
        this.head = new Node();
        this.tail = new Node();
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        Node search = this.cache.get(key);
        if(search == null){
            return -1;
        }else{
            moveToHead(search);
        }
        return search.value;
    }
    
    public void put(int key, int value) {
        Node search = this.cache.get(key);
        if(search == null){
            Node newnode = new Node();
            newnode.key = key;
            newnode.value = value;

            addNode(newnode);
            cache.put(key,newnode);
            this.curSize++;

            if(this.curSize > this.capacity){
                Node tail = deleteTail();
                this.cache.remove(tail.key);
                curSize--;
            }
        }else{
            search.value = value;
            moveToHead(search);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */


/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: python/146_LRU_Cache.py
from collections import OrderedDict
class LRUCache(OrderedDict):

    def __init__(self, capacity: int):
        self.capacity = capacity # self is a container of OrderedDict

    def get(self, key: int) -> int:
        if key not in self:
            return -1
        self.move_to_end(key,last=True) # move the recently visited one to the right end
        return self[key]

    def put(self, key: int, value: int) -> None:
        if key in self:
            self.move_to_end(key,last=True)# move the recently visited one to the right end
        self[key] = value
        if len(self) > self.capacity:
            self.popitem(last=False)# pop from left end

# dictionary + doulelinkedNode
class DLNode:
    def __init__(self):
        self.key = 0
        self.value = 0
        self.prev = None
        self.next = None
        
class LRUCache:
    def _add_node(self,node:DLNode) -> None:
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node
    
    def _remove_node(self,node:DLNode) -> None:
        pre= node.prev
        nxt = node.next
        pre.next = nxt
        nxt.prev = pre
    
    def _move_to_head(self,node:DLNode) -> None:
        self._remove_node(node)
        self._add_node(node)
    
    def _delete_tail(self) -> DLNode:
        real_tail = self.tail.prev
        self._remove_node(real_tail)
        return real_tail

    def __init__(self, capacity: int):
        self.cur_size = 0
        self.capacity = capacity
        self.cache = {}
        self.head = DLNode()
        self.tail = DLNode()
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        res = self.cache[key]
        self._move_to_head(res)
        return res.value;

    def put(self, key: int, value: int) -> None:
        if key not in self.cache:
            new_node = DLNode()
            new_node.key = key;
            new_node.value = value;
            
            self.cache[key] = new_node
            self.cur_size += 1
            self._add_node(new_node);
            
            if self.cur_size > self.capacity:
                tail = self._delete_tail()
                self.cur_size -= 1
                self.cache.pop(tail.key)
        else:
            res = self.cache[key]
            res.value = value
            self._move_to_head(res)
            
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)